File: src/api/routers/networks.py
from typing import Literal

from fastapi import APIRouter, Depends, Query

from src.api.dependencies import get_network_service
from src.api.schemas.networks import (
    CentralityLeadersResponse,
    NetworkResponse,
    TablePage,
)
from src.api.services.network_service import NetworkService

router = APIRouter(prefix="/runs", tags=["networks"])


@router.get("/{run_id}/network", response_model=NetworkResponse)
def get_network(
    run_id: str,
    metric: Literal["if", "wif"] = "if",
    community_id: str | None = None,
    period: str | None = Query(default=None, pattern=r"^\d{4}-(0[1-9]|1[0-2])$"),
    view: Literal["users", "communities"] = "users",
    sampling: (
        Literal["community_balanced", "strongest_edges", "full_graph"] | None
    ) = None,
    max_nodes: int = Query(default=200, ge=2),
    max_edges: int = Query(default=500, ge=1),
    min_weight: float = Query(default=0.0, ge=0.0),
    service: NetworkService = Depends(get_network_service),
) -> NetworkResponse:
    return NetworkResponse(
        **service.graph(
            run_id,
            metric=metric,
            community_id=community_id,
            min_weight=min_weight,
            max_nodes=max_nodes,
            max_edges=max_edges,
            period=period,
            view=view,
            sampling=sampling,
        )
    )


@router.get("/{run_id}/centrality", response_model=TablePage)
def get_centrality(
    run_id: str,
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    service: NetworkService = Depends(get_network_service),
) -> TablePage:
    return TablePage(**service.centrality(run_id, limit=limit, offset=offset))


@router.get(
    "/{run_id}/centrality-leaders",
    response_model=CentralityLeadersResponse,
)
def get_centrality_leaders(
    run_id: str,
    metric: Literal["if", "wif"] = "if",
    period: str | None = Query(default=None, pattern=r"^\d{4}-(0[1-9]|1[0-2])$"),
    service: NetworkService = Depends(get_network_service),
) -> CentralityLeadersResponse:
    return CentralityLeadersResponse(
        **service.centrality_leaders(run_id, metric=metric, period=period)
    )

File: frontend/src/api/networks.ts
import type {
  CentralityLeadersResponse,
  CommunitiesResponse,
  CommunityDetail,
  MetricName,
  NetworkResponse,
  NetworkView,
  SamplingStrategy,
  TablePage,
} from '../types/api';
import { getJson } from './client';
import { API_ROUTES, fillRoute } from './routes';

export interface NetworkParams {
  metric?: MetricName;
  period?: string;
  view?: NetworkView;
  sampling?: SamplingStrategy;
  community_id?: string;
  max_nodes?: number;
  max_edges?: number;
  min_weight?: number;
}

export interface PageParams {
  limit?: number;
  offset?: number;
}

export const getNetwork = (
  runId: string,
  params: NetworkParams = {},
  signal?: AbortSignal,
) =>
  getJson<NetworkResponse>(fillRoute(API_ROUTES.network, { run_id: runId }), {
    params,
    signal,
  });

export const getCentrality = (
  runId: string,
  params: PageParams = {},
  signal?: AbortSignal,
) =>
  getJson<TablePage>(fillRoute(API_ROUTES.centrality, { run_id: runId }), {
    params,
    signal,
  });

export const getCentralityLeaders = (
  runId: string,
  params: { metric?: MetricName; period?: string } = {},
  signal?: AbortSignal,
) =>
  getJson<CentralityLeadersResponse>(
    fillRoute(API_ROUTES.centralityLeaders, { run_id: runId }),
    { params, signal },
  );

export const getCommunities = (
  runId: string,
  params: PageParams & { metric?: MetricName; period?: string } = {},
  signal?: AbortSignal,
) =>
  getJson<CommunitiesResponse>(
    fillRoute(API_ROUTES.communities, { run_id: runId }),
    { params, signal },
  );

export const getCommunity = (
  runId: string,
  communityId: string,
  params: Omit<NetworkParams, 'community_id'> = {},
  signal?: AbortSignal,
) =>
  getJson<CommunityDetail>(
    fillRoute(API_ROUTES.communityDetail, {
      run_id: runId,
      community_id: communityId,
    }),
    { params, signal },
  );
